File: app/src/main/java/com/starsep/sokoban/database/HighScore.java
package com.starsep.sokoban.database;

import android.content.Context;

import com.starsep.sokoban.MainMenuActivity;

public class HighScore {
	public class DifferentLevelsException extends Exception {}

	public int hash;
	public int time;
	public int moves;
	public int pushes;

	public HighScore(int hash, int time, int moves, int pushes) {
		this.hash = hash;
		this.time = time;
		this.moves = moves;
		this.pushes = pushes;
	}

	public void improve(HighScore another) throws DifferentLevelsException {
		if (hash != another.hash) {
			throw new DifferentLevelsException();
		}
		time = Math.min(time, another.time);
		moves = Math.min(moves, another.moves);
		pushes = Math.min(pushes, another.pushes);
	}

	@Override
	public String toString() {
		return "HighScore(hash=" + hash + ", " +
				"time=" + time + ", " +
				"moves=" + moves + ", " +
				"pushes=" + pushes + ")";
	}
}

File: app/src/main/kotlin/com.starsep.sokoban.release/gamelogic/HighScore.kt
package com.starsep.sokoban.release.gamelogic

import androidx.room.Entity

@Entity(primaryKeys = [("levelNumber"), ("levelHash")])
data class HighScore(
        var levelNumber: Int,
        var time: Int = 0,
        var moves: Int = 0,
        var pushes: Int = 0,
        var levelHash: Int = 0
) {
    fun improve(another: HighScore) {
        time = Math.min(time, another.time)
        moves = Math.min(moves, another.moves)
        pushes = Math.min(pushes, another.pushes)
    }

    override fun equals(other: Any?): Boolean {
        if (other !is HighScore) {
            return false
        }
        return time == other.time &&
                moves == other.moves &&
                pushes == other.pushes
    }

    override fun toString(): String {
        return "HighScore(" +
                "time=" + time + ", " +
                "movesLive=" + moves + ", " +
                "pushes=" + pushes + ")"
    }

    override fun hashCode(): Int {
        return toString().hashCode()
    }
}
